### PR TITLE
[buildingplan] step4: generalize buildingplan to handle all building types

### PR DESCRIPTION
**Pull request by myk002**
#1640
but restrict to only the currently supported set so we can still assume only one filter is required for each building.

Code can be reviewed after #1674 is merged.

changes:
- update buildingplan plugin version to 2.0
- new serialization format for planned buildings
- old persistent data is automatically migrated to new format on load
- algorithm now respects job_item filters; items must match job_item filter and buildingplan ItemFilter
- more invalid items are now filtered out, like items encased in ice. are there any others we should be checking (see BadFlags struct)
- items are sorted before job is unsuspended so final item ordering is correct regardless of what order the items were matched and attached
- item counts in filters are kept up to date so if buildingplan is disabled before all filters are matched and the building is completed by DF itself, the item counts will come out correct (though the item ordering and building "roughness" may be incorrect)
- fixes two memory leaks in building finalization code
- allows artifacts to be matched (ItemFilter defaults now top out at Masterful -- Artifact is selectable but must be manually specified)
- add gui to switch between items for buildings that require multiple item types

### CODE REVIEW

File: plugins/buildingplan-planner.cpp [hunk 1]
```diff
@@ -1,531 +1,922 @@
...
 }
 
-void Planner::adjustMaxQuality(df::building_type type, int amount)
+static bool matchesFilters(df::item * item,
+                           df::job_item * job_item,
+                           const ItemFilter & item_filter)
 {
-    auto max_quality = &getDefaultItemFilterForType(type)->max_quality;
-    *max_quality = static_cast<df::item_quality>(*max_quality + amount);
+    if (job_item->item_type > -1 && job_item->item_type != item->getType())
```
[the remark this answers is not in the document]
Comment (lethosor) [reply plugins/buildingplan-planner.cpp:707]: Not sure I know enough to answer your question, sorry. It looks like https://github.com/DFHack/dfhack/blob/58917b03d167ba96afd5e5d66e7ee6ab7a4b0a20/library/modules/Items.cpp#L282 calls https://github.com/DFHack/dfhack/blob/58917b03d167ba96afd5e5d66e7ee6ab7a4b0a20/library/modules/Items.cpp#L261 by default (including when called from Jobs::isSuitableItem), and it's possible that the latter ends up checking item types implicitly.

**Review by lethosor — APPROVED** [review]
Looks good! I tested a couple single-item buildings (e.g. tables and chairs). Are there any building types with multiple items supported in this PR, or are those in #1676/others?

File: plugins/buildingplan-planner.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (myk002) [reply plugins/buildingplan-planner.cpp:707]: When I traced through the code, it seemed to me what was happening was that if any item in the vector was the appropriate type, then matches() would assume that *every* item in the vector was an acceptable type. This is a problem for any filter that doesn't have a vector_id set, like any of the tools. I was getting hives built directly out of bars of coal.

Adding the checks in buildingplan solves the issue for my use case, of course, but I can't help but think the are other code paths that would benefit from proper item matching.

**Comment by myk002** [comment]
> Are there any building types with multiple items supported in this PR, or are those in #1676/others?

No, this PR doesn't directly add support for any new building types. All changes are "under the hood"

File: plugins/buildingplan-planner.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (myk002) [reply plugins/buildingplan-planner.cpp:707]: I was going to say that I bet advfort has this bug, but I see that isSuitableItem is locally implemented for that script (though the library isSuitableItem() and advfort's isSuitableItem() check different flags). In fact, the library implementation looks like it should have a lot of what advfort's isSuitableItem() function does.

[the remark this answers is not in the document]
Comment (lethosor) [reply plugins/buildingplan-planner.cpp:708]: The main difference I see from a glance is that advfort's `isSuitableItem` takes the item, and `Jobs::isSuitableItem` just takes an item type+subtype, so advfort's has the ability to call some more item methods. I'm not sure if `Jobs::isSuitableItem` is intentionally limited or if it was just written under the assumption that the item type+subtype would be enough.

I'm not very familiar with some of the code in these modules, so I'll defer to you if you think improvements are in order, but I think the checks you added make sense where you added them (at least given the current API).

**Pull request merged**